### PR TITLE
Add configurable reasoner plan logging & periodic printing of counters 

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## What is the goal of this PR?
Adds logging of reasoner plans and reasoner performance counters. These can be toggled through the config file &  log-level of respective loggers.

To enable & log reasoner plans:
```
log:
  ...
  logger:
    ...
    reasoner-planner:
        filter: com.vaticle.typedb.core.reasoner.planner
        level: debug
        output: [ stdout, file ]

```

To enable & log reasoner counters:
```
log:
  ...
  logger:
    ...
    reasoner-perf-counters:
      filter: com.vaticle.typedb.core.reasoner.common.ReasonerPerfCounters
      level: debug
      output: [ stdout ]
  ...
  debugger:
    reasoner-tracer:
      enable: false
      type: reasoner-tracer
      output: file
    reasoner-perf-counters:
      enable: true
      type: reasoner-perf-counters        
```

## What are the changes implemented in this PR?
* Adds periodic logging of reasoner perf counters when 'enabled' in the config file
* Adds logging of reasoner plans at debug level through the logger `com.vaticle.typedb.core.reasoner.planner.ReasonerPlanner`
* Adds a `RetrievableProcessor` counter to the reasoner performance counters.

### CODE REVIEW

**Comment by krishnangovindraj** [comment]
Sample output:
```

================================================= Start reasoner plans =================================================
--------------------------------		<user>::[]		--------------------------------
Cost: 58514.0
[0] RET {}
	{ $di-id isa id; $di-id == "C0002874";  $di has $di-id; $di isa disease; }
[1] RET {}
	{ $dr-id isa id; $dr-id == "chembl:CHEMBL2105680";  $dr isa drug; $dr has $dr-id; }
[2] CON {$di, $dr}
	{ $x isa potential-disease-treatment; $x ($di, $dr); }
		- inferred-potential-disease-treatment::[$di, $dr]

========================================================================================================================
--------------------------------		inferred-potential-disease-treatment::[$di, $dr]		--------------------------------
Cost: 1.9206702396636E13
[0] RET {$dr}
	{ $dr isa drug; }
[1] RET {$di}
	{ $di isa disease; }
[2] CON {$di}
	{ $_2 (associated-protein:$p, associated-disease:$di); $_2 isa disease-protein-association; }
		- inferred-disease-protein-association::[$d]
[3] RET {$p}
	{ $p isa protein; }
[4] CON {$p, $dr}
	{ $_1 (associated-drug:$dr, associated-protein:$p); $_1 isa drug-protein-association; }
		- inferred-drug-protein-association::[$p, $d]

========================================================================================================================
--------------------------------		inferred-drug-protein-association::[$p, $d]		--------------------------------
Cost: 2.94782723E8
[0] RET {$d}
	{ $d isa drug; }
[1] RET {$p}
	{ $p isa protein; }
[2] CON {$p, $d}
	{ $_1 isa drug-protein-interaction; $_1 (interacting-drug:$d, interacting-protein:$p); }
		- inferred-drug-protein-interaction::[$p, $d]
- - - - - - - - - - - - - - - - - - - - - - - - NEXT BRANCH  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Cost: 5.393581289E9
[0] RET {$d}
	{ $d isa drug; }
[1] RET {$p}
	{ $p isa protein; }
[2] CON {$d}
	{ $_1 isa drug-protein-interaction; $_1 (interacting-drug:$d, interacting-protein:$p2); }
		- inferred-drug-protein-interaction::[$d]
[3] CON {$p, $p2}
	{ $_2 isa protein-protein-association; $_2 (associated-protein:$p2, associated-protein:$p); }
		- inferred-protein-protein-association::[$p2, $p1]
		- inferred-protein-protein-association-2::[$p2, $p1]

========================================================================================================================
--------------------------------		inferred-disease-protein-association::[$d]		--------------------------------
Cost: 1.809669172E9
[0] RET {$d}
	{ $d isa disease; }
[1] CON {$d}
	{ $_1 isa disease-protein-interaction; $_1 (interacting-disease:$d, interacting-protein:$p2); }
		- inferred-disease-protein-interaction::[$d]
[2] CON {$p2}
	{ $_2 isa protein-protein-association; $_2 (associated-protein:$p2, associated-protein:$p); }
		- inferred-protein-protein-association::[$p1]
		- inferred-protein-protein-association::[$p2]
		- inferred-protein-protein-association-2::[$p2]
		- inferred-protein-protein-association-2::[$p1]
[3] RET {$p}
	{ $p isa protein; }
- - - - - - - - - - - - - - - - - - - - - - - - NEXT BRANCH  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Cost: 1.4673461E7
[0] RET {$d}
	{ $d isa disease; }
[1] CON {$d}
	{ $_1 isa disease-protein-interaction; $_1 (interacting-disease:$d, interacting-protein:$p); }
		- inferred-disease-protein-interaction::[$d]
[2] RET {$p}
	{ $p isa protein; }

========================================================================================================================
--------------------------------		inferred-protein-protein-association::[$p2, $p1]		--------------------------------
Cost: 2.265858537E9
[0] NEG {$p2, $p1}
{	Cost: 2.0
	[0] RET {$p2, $p1}
		{ $p1 is $p2; }
}
[1] CON {$p2}
	{ $_2 isa protein-encoding; $_2 (encoding-gene:$g2, encoded-protein:$p2); }
		- inferred-protein-encoding::[$p]
[2] CON {$p1}
	{ $_1 isa protein-encoding; $_1 (encoding-gene:$g1, encoded-protein:$p1); }
		- inferred-protein-encoding::[$p]
[3] RET {$g1, $g2}
	{ $_3 (encoding-gene:$g1, encoding-gene:$g2); $_3 isa encoded-protein-interaction;  $g1 isa gene;  $g2 isa gene; }

========================================================================================================================
--------------------------------		inferred-drug-protein-interaction::[$d]		--------------------------------
Cost: 550006.0
[0] RET {$d}
	{ $_1 isa drug-gene-interaction; $_1 (interacting-drug:$d, interacting-gene:$g);  $d isa drug;  $g isa gene; }
[1] CON {$g}
	{ $_2 isa protein-encoding; $_2 (encoding-gene:$g, encoded-protein:$p); }
		- inferred-protein-encoding::[$g]
[2] RET {$p}
	{ $p isa protein; }

========================================================================================================================
--------------------------------		inferred-protein-protein-association::[$p1]		--------------------------------
Cost: 4016818.0
[0] CON {$p1}
	{ $_1 isa protein-encoding; $_1 (encoding-gene:$g1, encoded-protein:$p1); }
		- inferred-protein-encoding::[$p]
[1] RET {$g1}
	{ $_3 (encoding-gene:$g1, encoding-gene:$g2); $_3 isa encoded-protein-interaction;  $g1 isa gene;  $g2 isa gene; }
[2] CON {$g2}
	{ $_2 isa protein-encoding; $_2 (encoding-gene:$g2, encoded-protein:$p2); }
		- inferred-protein-encoding::[$g]
[3] NEG {$p2, $p1}
{	Cost: 2.0
	[0] RET {$p2, $p1}
		{ $p1 is $p2; }
}

========================================================================================================================
--------------------------------		inferred-protein-protein-association::[$p2]		--------------------------------
Cost: 4016818.0
[0] CON {$p2}
	{ $_2 isa protein-encoding; $_2 (encoding-gene:$g2, encoded-protein:$p2); }
		- inferred-protein-encoding::[$p]
[1] RET {$g2}
	{ $_3 (encoding-gene:$g1, encoding-gene:$g2); $_3 isa encoded-protein-interaction;  $g1 isa gene;  $g2 isa gene; }
[2] CON {$g1}
	{ $_1 isa protein-encoding; $_1 (encoding-gene:$g1, encoded-protein:$p1); }
		- inferred-protein-encoding::[$g]
[3] NEG {$p2, $p1}
{	Cost: 2.0
	[0] RET {$p2, $p1}
		{ $p1 is $p2; }
}

========================================================================================================================
--------------------------------		inferred-protein-protein-association-2::[$p2, $p1]		--------------------------------
Cost: 1.1577843E7
[0] NEG {$p2, $p1}
{	Cost: 2.0
	[0] RET {$p2, $p1}
		{ $p1 is $p2; }
}
[1] RET {$p2, $p1}
	{ $_2 (participating-protein:$p2, participated-pathway:$path); $_2 isa pathway-participation;  $_1 (participating-protein:$p1, participated-pathway:$path); $_1 isa pathway-participation; }

========================================================================================================================
--------------------------------		inferred-drug-protein-interaction::[$p, $d]		--------------------------------
Cost: 424477.0
[0] RET {$p}
	{ $p isa protein; }
[1] CON {$p}
	{ $_2 isa protein-encoding; $_2 (encoding-gene:$g, encoded-protein:$p); }
		- inferred-protein-encoding::[$p]
[2] RET {$d, $g}
	{ $_1 isa drug-gene-interaction; $_1 (interacting-drug:$d, interacting-gene:$g);  $d isa drug;  $g isa gene; }

========================================================================================================================
--------------------------------		inferred-disease-protein-interaction::[$d]		--------------------------------
Cost: 7188648.0
[0] RET {$d}
	{ $_2 isa disease-gene-interaction; $_2 (interacting-gene:$g, interacting-disease:$d);  $d isa disease;  $g isa gene; }
[1] CON {$g}
	{ $_1 isa protein-encoding; $_1 (encoded-protein:$p, encoding-gene:$g); }
		- inferred-protein-encoding::[$g]
[2] RET {$p}
	{ $p isa protein; }

========================================================================================================================
--------------------------------		inferred-protein-protein-association-2::[$p2]		--------------------------------
Cost: 1.7366762E7
[0] RET {$p2}
	{ $_2 (participating-protein:$p2, participated-pathway:$path); $_2 isa pathway-participation;  $_1 (participating-protein:$p1, participated-pathway:$path); $_1 isa pathway-participation; }
[1] NEG {$p2, $p1}
{	Cost: 2.0
	[0] RET {$p2, $p1}
		{ $p1 is $p2; }
}

========================================================================================================================
--------------------------------		inferred-protein-protein-association-2::[$p1]		--------------------------------
Cost: 1.7366762E7
[0] RET {$p1}
	{ $_2 (participating-protein:$p2, participated-pathway:$path); $_2 isa pathway-participation;  $_1 (participating-protein:$p1, participated-pathway:$path); $_1 isa pathway-participation; }
[1] NEG {$p2, $p1}
{	Cost: 2.0
	[0] RET {$p2, $p1}
		{ $p1 is $p2; }
}

========================================================================================================================
--------------------------------		inferred-protein-encoding::[$g]		--------------------------------
Cost: 95157.0
[0] RET {$g}
	{ $p isa protein;  $_2 (translated-transcript:$t, synthesised-protein:$p); $_2 isa translation;  $_1 (transcribed-gene:$g, synthesised-transcript:$t); $_1 isa transcription;  $t isa transcript;  $g isa gene; }

========================================================================================================================
--------------------------------		inferred-protein-encoding::[$p]		--------------------------------
Cost: 95157.0
[0] RET {$p}
	{ $p isa protein;  $_2 (translated-transcript:$t, synthesised-protein:$p); $_2 isa translation;  $_1 (transcribed-gene:$g, synthesised-transcript:$t); $_1 isa transcription;  $t isa transcript;  $g isa gene; }

========================================================================================================================
=================================================  End reasoner plans  =================================================

```

File: reasoner/planner/ReasonerPlanner.java [hunk 1]
```diff
@@ -72,15 +81,27 @@ static boolean dependenciesSatisfied(Resolvable<?> resolvable, Set<Variable> bou
...
-                .forEach(callMode -> plan(callMode));
+    public void planRoot(ResolvableConjunction conjunction) {
+        long start = System.nanoTime();
+        CallMode callMode = new CallMode(conjunction, Collections.emptySet());
+        plan(callMode);
+        perfCounters.timePlanning.add(System.nanoTime() - start);
+        logPlans(Iterators.single(callMode));
+    }
+
+    public void planExplainableRoot(Concludable concludable, Set<Variable> mode) {
```
Comment (flyingsilverfin) [inline reasoner/planner/ReasonerPlanner.java:99]: Mode is unused - you should listen to intellij's feedback!

File: reasoner/planner/ReasonerPlanner.java [hunk 2]
```diff
@@ -177,6 +198,16 @@ public Set<CallMode> triggeredCalls(Concludable concludable, Set<Variable> mode,
         return calls;
     }
 
+    private void logPlans(FunctionalIterator<CallMode> roots) {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Printing reasoner plans:\n" + ReadablePlan.prettyString(summarisePlans(roots.toSet())));
+        }
+    }
+
+    public Set<ReadablePlan> summarisePlans(Set<CallMode> roots) {
```
Comment (flyingsilverfin) [inline reasoner/planner/ReasonerPlanner.java:207]: Why is this public? And it's only used in 1 place?

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]
I've just done a bit of a review

File: reasoner/planner/ReasonerPlanPrinter.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                sb.append(nesting).append("\t\t- ").append(ruleMode.label).append("::").append(ruleMode.mode).append("\n");
+            });
+        }
+    }
+
+    private void appendRuleModeHeader(RuleMode ruleMode) {
+        sb.append("--------------------------------\t\t").append(ruleMode.label).append("::").append(ruleMode.mode).append("\t\t--------------------------------\n");
+    }
+
+    private void appendResolvableHeader(String nesting, int resolvableIndex, String resolvableType, Set<Variable> bounds) {
```
Comment (flyingsilverfin) [inline reasoner/planner/ReasonerPlanPrinter.java:122]: is `resolvableIndex` the same as `planIndex`?

File: reasoner/planner/ReasonerPlanner.java [hunk 4]
```diff
@@ -177,6 +196,12 @@ public Set<CallMode> triggeredCalls(Concludable concludable, Set<Variable> mode,
         return calls;
     }
 
+    private void logPlans(Set<ResolvableConjunction> roots, Set<Variable> mode) {
```
Comment (flyingsilverfin) [inline reasoner/planner/ReasonerPlanner.java:199]: I'd call this `mayLogPlans` since it doesn't always log

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]
Left some single comments

File: reasoner/planner/ReasonerPlanPrinter.java (hunk 3 above)
Comment (flyingsilverfin) [reply reasoner/planner/ReasonerPlanPrinter.java:122]: @krishnangovindraj  this one too please

File: reasoner/planner/ReasonerPlanPrinter.java [hunk 5]
```diff
@@ -147,18 +147,18 @@ private Set<RuleMode> triggeredRuleModes(Concludable concludable, Set<Variable>
...
-        Set<RuleMode> triggeredRuleModes = new HashSet<>();
-        builder.forEach((labelMode, conjunctions) -> triggeredRuleModes.add(new RuleMode(labelMode.first(), labelMode.second(), conjunctions)));
-        return triggeredRuleModes;
+        Set<LabelGroup> triggeredLabelGroups = new HashSet<>();
+        builder.forEach((labelMode, conjunctions) -> triggeredLabelGroups.add(new LabelGroup(labelMode.first(), labelMode.second(), conjunctions)));
+        return triggeredLabelGroups;
     }
 
-    private static class RuleMode {
+    private static class LabelGroup {
```
Comment (flyingsilverfin) [inline reasoner/planner/ReasonerPlanPrinter.java:155]: Hmm I'd expect a `LabelGroup` to contain a group of Labels. Here it appears there's one group of conjunctions for one label/mode?
What about `CallGroup`?

File: reasoner/planner/ReasonerPlanPrinter.java (hunk 3 above)
Comment (krishnangovindraj) [reply reasoner/planner/ReasonerPlanPrinter.java:122]: It's the index of the resolvable in the plan (resolvable ordering) .

File: reasoner/planner/ReasonerPlanner.java [hunk 6]
```diff
@@ -177,6 +198,16 @@ public Set<CallMode> triggeredCalls(Concludable concludable, Set<Variable> mode,
         return calls;
     }
 
+    private void logPlans(FunctionalIterator<CallMode> roots) {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Printing reasoner plans:\n" + ReadablePlan.prettyString(summarisePlans(roots.toSet())));
```
[the remark this answers is not in the document]
Comment (krishnangovindraj) [reply reasoner/planner/ReasonerPlanner.java]: There's too much work being done to hide away in `toString`. I'd rather name it something explicit. Maybe serialise?

**Review by krishnangovindraj — COMMENTED** [review]
replies & updates

**Pull request merged**